### PR TITLE
Fix notifications appearing during gameplay

### LINKED ISSUE #2643: Can't open overlays at osu! logo
Should still be able to open overlays here.

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/58860112-can-t-open-overlays-at-osu-logo?utm_campaign=plugin&utm_content=tracker%2F4707269&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F4707269&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### LINKED ISSUE #2649: Simplify overlay visibility logic
Because of recurring issues and increasing demands to the functionality (#2613 #2640) in `Visibility` for overlays (basically everything accessible through the toolbar including the toolbar itself) it was proposed to change the way this works. Instead of two `Bindables` (currently `AllowOpeningOverlays` and `HideOverlaysOnEnter`) a single `Bindable` enum is suggested.

`HideOverlaysOnEnter` should remain as a `virtual bool`.

```csharp
enum OverlayActivation
{
    Disabled,
    UserTriggered,
    All
}
```
While `Disabled` and `UserTriggered` should be self-explanatory, `All` is the only way the game is allowed to interact.

Currently there is no way to discern if the `Visibility` was changed through user interaction so this potentially requires a framework change.

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/58888388-simplify-overlay-visibility-logic?utm_campaign=plugin&utm_content=tracker%2F4707269&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F4707269&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by Aergwyn**
+ Resolves #2649
+ fix `Toolbar` being toggleable when it shouldn't be able to, e.g. `Intro` (can't find the issue D:)
+ allow opening overlays in `MenuState.Initial` again (resolves #2643)

This isn't *complete* as it's missing `OverlayActivation.UserTriggered` but I'm not sure how to do that yet and I thought that this PR could either be used
a) for feedback how to accomplish that or
b) may be valid before adding the last bit of functionality.

### CODE REVIEW

File: osu.Game/Graphics/Containers/OsuFocusedOverlayContainer.cs [hunk 1]
```diff
@@ -16,13 +17,16 @@ public class OsuFocusedOverlayContainer : FocusedOverlayContainer
         private SampleChannel samplePopIn;
         private SampleChannel samplePopOut;
 
-        private readonly BindableBool allowOpeningOverlays = new BindableBool(true);
+        /// <summary>
+        /// Defaults to <see cref="OverlayActivation.All"/> so that the overlay works even if BDL couldn't pass an <see cref="OsuGame"/>.
```
Comment (peppy) [inline osu.Game/Graphics/Containers/OsuFocusedOverlayContainer.cs]: I don't think the default needs to be explained.

File: osu.Game/Screens/Menu/ButtonSystem.cs [hunk 2]
```diff
@@ -332,19 +329,19 @@ private void updateLogoState(MenuState lastState = MenuState.Initial)
                 case MenuState.Exit:
                 case MenuState.Initial:
                     logoTracking = false;
+                    allowOverlays.Value = OverlayActivation.Disabled;
```
Comment (peppy) [inline osu.Game/Screens/Menu/ButtonSystem.cs]: I'm pretty sure we don't want this `Disabled`. If anything, I'd uncomment `UserTriggered` and set that here so we aren't outright disabling for the time being (even if the user triggered support is not in yet).

Comment (peppy) [reply osu.Game/Screens/Menu/ButtonSystem.cs]: Also this is probably a use-case for hiding overlays *once* via a method call, but not actually disabling. Without the `All` set below in the delayed delegate (which is bound to cause weirdness).

File: osu.Game/Screens/OsuScreen.cs [hunk 3]
```diff
@@ -103,8 +109,18 @@ private void load(OsuGameBase game, OsuGame osuGame, AudioManager audio)
             if (osuGame != null)
             {
                 Ruleset.BindTo(osuGame.Ruleset);
-                hideOverlaysOnEnter.BindTo(osuGame.HideOverlaysOnEnter);
-                allowOpeningOverlays.BindTo(osuGame.AllowOpeningOverlays);
+                AllowOverlays.BindTo(osuGame.AllowOverlays);
```
Comment (peppy) [inline osu.Game/Screens/OsuScreen.cs]: This is seriously dangerous. You're both binding to the global bindable and changing the value of it in BDL async methods. This means if a screen is preloaded without being immediately pushed, it is potentially changing the state of this variable.

**Review by peppy — CHANGES_REQUESTED** [review]

File: osu.Game/Screens/Menu/ButtonSystem.cs (hunk 2 above)
Comment (Aergwyn) [reply osu.Game/Screens/Menu/ButtonSystem.cs]: Regarding the first comment: It completely falls flat as I noticed that any logic for that in `ButtonSystem` is now unnecessary.
Regarding the second comment: Every time the `MenuState` changes the visibility of the overlays is reevaluated as seen [here](https://github.com/Aergwyn/osu/blob/707af0209732ef29de79d4417c2aab9892ff8f89/osu.Game/Screens/Menu/MainMenu.cs#L67).

File: osu.Game/Screens/Menu/MainMenu.cs [hunk 4]
```diff
@@ -64,6 +63,8 @@ public MainMenu()
                 },
                 sideFlashes = new MenuSideFlashes(),
             };
+
+            buttons.StateChanged += state => UpdateOverlayStates?.Invoke();
```
Comment (peppy) [inline osu.Game/Screens/Menu/MainMenu.cs]: this can't be

File: osu.Game/Screens/Menu/ButtonSystem.cs [hunk 5]
```diff
@@ -331,9 +321,6 @@ private void updateLogoState(MenuState lastState = MenuState.Initial)
                     {
                         logoTracking = false;
 
-                        hideOverlaysOnEnter.Value = true;
```
Comment (peppy) [inline osu.Game/Screens/Menu/ButtonSystem.cs]: moving this out has caused a regression in the menu animation (timings have changed)

File: osu.Game/Screens/OsuScreen.cs [hunk 6]
```diff
@@ -103,8 +110,18 @@ private void load(OsuGameBase game, OsuGame osuGame, AudioManager audio)
...
-                hideOverlaysOnEnter.BindTo(osuGame.HideOverlaysOnEnter);
-                allowOpeningOverlays.BindTo(osuGame.AllowOpeningOverlays);
+                allowOverlays.BindTo(osuGame.AllowOverlays);
+
+                updateOverlayStates = () =>
+                {
+                    if (HideOverlaysOnEnter)
+                    {
+                        osuGame.CloseAllOverlays();
+                        osuGame.Toolbar.State = Visibility.Hidden;
```
Comment (peppy) [inline osu.Game/Screens/OsuScreen.cs]: why is the toolbar state handled here rather than in `OsuGame`?

**Review by peppy — CHANGES_REQUESTED** [review]

**Review by peppy — APPROVED** [review]

**Pull request merged**